Fixes for vdw and fmt libraries

mtaillefumier opened the pull request:
This PR includes major fixes:

- vdw MPI issue is finally fixed. MPI_all_reduce was missing when calculating the energy contribution of the vdw functional
- Added a message indicating that C_LDA correction is needed when libvdwxc functional are used. 
- XC functionals can be weighted. It is set to one by default
- Compared SIRIUS+libvdwxc to native qe:
    -  Difference in total energy < 1 mHa 
    -  Forces are also correct
    - Stress tensor difference between QE and SIRIUS below 1kbar
- fmt header also fixed

Commit:
[vdw] Final revisions for vdw

- compared SIRIUS implementation of vdw functionals with q-e native.
  - Difference in energy below the mHa
  - No forces originating from vdw functionals
  - stress tensor compared to QE implementation. Difference below 1 kbar
- Added the posibility to reweight each functional.
  By default all functionals have equal weight.
- vdw MPI issue solved. We need a mpi_all_reduce to compute
  the energy of the non-lcal vdw correction.
- Fix compilation errors with fmt
-

Update CMakeLists.txt

fix an error

Update xc.cpp

unnecessary formatting

Object type has no default values

Fix dftd3 and dftd4 stress tensors. dftd3 and dftd4 use atomic units

Added vdw stress tensor glue code

More testing

Commit:
Fix compilation errors with fmt

mtaillefumier commented:
cscs-ci run default

simonpintarelli committed:
Update k_point_set.cpp

simonpintarelli commented:
Thanks @mtaillefumier!

simonpintarelli reviewed (approved):
lgtm

simonpintarelli commented:
cscs-ci run default

simonpintarelli commented:
daint gpu band/k-point parallel tests are failing in test35
```
204: total energy is different: -158.617337 computed vs. -158.616928 reference
```

https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/7415069138667150/4626796664769983/-/jobs/10627183108#L2653

On eiger it runs into walltime.

Ref: https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/7415069138667150/4626796664769983/-/pipelines/1917097357

simonpintarelli commented:
cscs-ci run default

mtaillefumier commented:
> test35 gave
> 
> ```
> 204: total energy is different: -158.617337 computed vs. -158.616928 reference
> ```

I also saw it. I still can not figure out why including a header file would change the result though.

simonpintarelli commented:
> > test35 gave
> > ```
> > 204: total energy is different: -158.617337 computed vs. -158.616928 reference
> > ```
> 
> I also saw it. I still can not figure out why including a header file would change the result though.

I don't think the header file did it ..

mtaillefumier commented:
I also have a segfault on mi300 which I do not fully understand yet either.

mtaillefumier commented:
cscs-ci run default

simonpintarelli commented:
cscs-ci run default

mtaillefumier commented:
The results differ between MPI=1 and MPI \neq 1. It is coming from the parallel fft and some strange interqaction between vdwxc and sirius slab decomposition.

mtaillefumier commented:
cscs-ci run default

mtaillefumier commented:
cscs-ci run default

mtaillefumier commented:
cscs-ci run default

toxa81 commented on src/context/simulation_context.cpp:
```diff
@@ -690,20 +691,32 @@ Simulation_context::print_info(std::ostream& out__) const
+
+        /* Check if the weights are given and if the number of elements match the number of functionals. If empty, then default to 1.0
+         */
+        if (xc_functionals_weight().size() != 0) {
```
Is `if (xc_functionals_weight().size() != 0) {` check redundant?

mtaillefumier replied:
Actually this one is needed if we want to have a default behavior. SIRIUS will throw an error only if the weights are given and the number of weight does not match the number of functional. If the weights are not explicitly given then we default to 1.0 for each functional.

toxa81 replied:
clear

toxa81 commented on src/context/simulation_parameters.hpp:
```diff
@@ -503,6 +503,12 @@ class Simulation_parameters
         return cfg().parameters().xc_functionals();
     }
 
+    std::vector<double>
```
can be `inline auto`

mtaillefumier replied:
Ok let me try this.

mtaillefumier replied:
Also included now. Thanks for the suggestion

toxa81 commented on src/potential/potential.cpp:
```diff
@@ -65,12 +65,27 @@ Potential::Potential(Simulation_context& ctx__)
+
+    /* Check if the weights are given and if the number of elements match the number of functionals. If empty, then default to 1.0
+     */
+    if (ctx_.xc_functionals_weight().size() != 0) {
```
is `if (ctx_.xc_functionals_weight().size() != 0) {` check redundant?

mtaillefumier replied:
See comment above. I can maybe try to merge these checks in one check only.

toxa81 replied:
yes, probably can be fused into one `if` using C's short-circuit evaluation of `if` statements

mtaillefumier replied:
Sorry I forgot to include this change. It is done now

mtaillefumier commented:
There is a bug in the API as well. When a field is declared object, SIRIUS still wants to have the "default" field defined which results in throwing an error.

simonpintarelli commented:
cscs-ci run default

simonpintarelli commented:
cscs-ci run default

simonpintarelli commented:
cscs-ci run default

toxa81 committed:
use inline auto

toxa81 committed:
simplify total force calculation

toxa81 commented on src/geometry/stress.cpp:
```diff
@@ -362,8 +380,43 @@ Stress::calc_stress_xc()
-    symmetrize_stress_tensor(ctx_.unit_cell().symmetry(), stress_xc_);
+    // compute the contributions of the stress tensor originating from non-local corrections
+    if (ctx_.full_potential()) {
+        potential_.xc_vdw_stress<false>(density_);
```
There is no stress code for full-potential. Maybe the switch is for non-local core correction?

mtaillefumier replied:
This is correct. I forgot about it. I should only call it when we use pseudo-potentials. I will fix it ASAP

mtaillefumier replied:
the template parameter is the same than in
```
 template <bool add_pseudo_core__>
void
Potential::xc(Density const& density__)
```

toxa81 replied:
you probably just need to keep `potential_.xc_vdw_stress<true>(density_);`

toxa81 reviewed (changes requested):
I found one suspicious place for stress tensor (full-potential switch)

simonpintarelli commented:
cscs-ci run default

Commit:
Stress tensor for full potential is not supported

Commit:
Added spack recipe for libvdwxc 0.5.0

mtaillefumier commented:
cscs-ci run default

Commit:
Fuse if statements

toxa81 committed:
Merge remote-tracking branch 'origin/develop' into fmt_fixes

simonpintarelli committed:
[fix] cmake consistent with nvtx3 update (#1082)

* require cmake@3.25

Commit:
test

Commit:
bug fix

Commit:
Update docker image

Commit:
[cmake] Bump mandatory libvdwxc version to 0.5.0

toxa81 commented:
hmm.. in the base container it's vdw-0.5.0.. why build_and_test_cpu finds 0.4.0

mtaillefumier commented:
Maybe I did a mistake when updating the docker file. I will check it

toxa81 commented:
I checked your build, it was actually fine. I triggered it once again https://github.com/electronic-structure/SIRIUS/actions/runs/20426901540/job/58689018912 but I don't think it will do something different

mtaillefumier commented:
compilation passes. I only need to update the last two tests as dftd3 and dftd4 stress tensors were changed.

Commit:
Updated dftd3 and dftd4 tests

Commit:
[dftdx] change sign of stress tensor. QE convention

Commit:
Test updates

toxa81 commented on src/geometry/stress.cpp:
```diff
@@ -362,8 +380,43 @@ Stress::calc_stress_xc()
 
-    symmetrize_stress_tensor(ctx_.unit_cell().symmetry(), stress_xc_);
+    // compute the contributions of the stress tensor originating from non-local corrections
+    if (ctx_.full_potential()) {
```
It can't be `if (ctx_.full_potential())`. Stress tensor only implemented for the pseudo-potential.

mtaillefumier commented:
@toxa81 I can not find the last change request. Would it be possible to check again so that I include these changes before merging.

Commit:
Stress tensor are only calculated for pseudo-potentials

toxa81 commented:
cscs-ci run default

toxa81 reviewed (approved):
LGTM

simonpintarelli committed:
typo

simonpintarelli commented:
Note: I'll rebase / cherry pick once the CI has been fixed the other PR (1113)

simonpintarelli committed:
use libvdwxc@develop (@0.5 -> broken)

libvdwxc@0.5+mpi doesn't work, the autconf script is broken and chokes on MPI detection.

simonpintarelli committed:
use libvdwxc@develop

mtaillefumier committed:
Merge branch 'develop' into fmt_fixes
